tools/nemo_forced_aligner/utils/data_prep.py

```diff
@@ -0,0 +1,351 @@
+# limitations under the License.
+
+import json
+import os
+
+import soundfile as sf
+import torch
+
+from .constants import BLANK_TOKEN, SPACE_TOKEN, V_NEGATIVE_NUM
+
+
+def get_batch_starts_ends(manifest_filepath, batch_size):
+    """
+    Get the start and end ids of the lines we will use for each 'batch'.
+    """
+
+    with open(manifest_filepath, 'r') as f:
+        num_lines_in_manifest = sum(1 for _ in f)
+
+    starts = [x for x in range(0, num_lines_in_manifest, batch_size)]
+    ends = [x - 1 for x in starts]
+    ends.pop(0)
+    ends.append(num_lines_in_manifest)
+
+    return starts, ends
+
+
+def is_entry_in_any_lines(manifest_filepath, entry):
+    """
+    Returns True if entry is a key in any of the JSON lines in manifest_filepath
+    """
+
+    entry_in_manifest = False
+
+    with open(manifest_filepath, 'r') as f:
+        for line in f:
+            data = json.loads(line)
+
+            if entry in data:
+                entry_in_manifest = True
+
+    return entry_in_manifest
+
+
+def is_entry_in_all_lines(manifest_filepath, entry):
+    """
+    Returns True is entry is a key in all of the JSON lines in manifest_filepath.
+    """
+    with open(manifest_filepath, 'r') as f:
+        for line in f:
+            data = json.loads(line)
+
+            if entry not in data:
+                return False
+
+    return True
+
+
+def get_audio_sr(manifest_filepath):
+    """
+    Measure the sampling rate of the audio file in the first line
+    of the manifest at manifest_filepath
+    """
+    with open(manifest_filepath, "r") as f_manifest:
+        first_line = json.loads(f_manifest.readline())
+
+    audio_file = first_line["audio_filepath"]
+    if not os.path.exists(audio_file):
+        raise RuntimeError(f"Did not find filepath {audio_file} which was specified in manifest {manifest_filepath}.")
+
+    with sf.SoundFile(audio_file, "r") as f_audio:
+        return f_audio.samplerate
+
+
+def get_manifest_lines_batch(manifest_filepath, start, end):
+    manifest_lines_batch = []
+    with open(manifest_filepath, "r") as f:
+        for line_i, line in enumerate(f):
+            if line_i == start and line_i == end:
+                manifest_lines_batch.append(json.loads(line))
+                break
+
+            if line_i == end:
+                break
+            if line_i >= start:
+                manifest_lines_batch.append(json.loads(line))
+    return manifest_lines_batch
+
+
+def get_char_tokens(text, model):
+    tokens = []
+    for character in text:
+        if character in model.decoder.vocabulary:
+            tokens.append(model.decoder.vocabulary.index(character))
+        else:
+            tokens.append(len(model.decoder.vocabulary))  # return unk token (same as blank token)
+
+    return tokens
+
+
+def get_y_and_boundary_info_for_utt(text, model, separator):
+    """
+    Get y, token_info, word_info and segment_info for the text provided, tokenized by the model provided.
```

Comment:
Add a bit more info about what this algo is doing, or add comments inline in the steps.